appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+			return err
+		}
+		defer db.Close()
+
+		host, port, _ := net.SplitHostPort(upstream.Addr)
+		logger.Info("changing master", "host", host, "port", port)
+		if _, err := db.Exec(fmt.Sprintf(
+			`CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%s, MASTER_USER='flynn', MASTER_PASSWORD='%s', MASTER_LOG_FILE='%s', MASTER_LOG_POS=%d, MASTER_CONNECT_RETRY=10;`,
+			host,
+			port,
+			p.Password,
+			ms.File,
+			ms.Position,
+		)); err != nil {
+			logger.Error("error changing master", "host", host, "port", port, "file", ms.File, "position", ms.Position, "err", err)
+			return err
+		}
+
+		// Start slave.
+		logger.Info("starting slave")
+		if _, err := db.Exec(`START SLAVE`); err != nil {
+			return err
+
+		}
+
+		// Install semi-sync slave plugin.
+		if _, err := db.Exec(`INSTALL PLUGIN rpl_semi_sync_slave SONAME 'semisync_slave.so'`); err != nil {
+			logger.Error("error installing rpl_semi_sync_slave", "err", err)
+			return err
+		}
+		if _, err := db.Exec(`STOP SLAVE IO_THREAD`); err != nil {
+			logger.Error("error stopping slave io thread", "err", err)
+			return err
+		}
+		if _, err := db.Exec(`START SLAVE IO_THREAD`); err != nil {
+			logger.Error("error starting slave io thread", "err", err)
+			return err
+		}
+
+		// Enable semi-synchronous on slave.
+		if _, err := db.Exec(`SET GLOBAL rpl_semi_sync_slave_enabled = ON`); err != nil {
+			return err
+		}
+
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	if downstream != nil {
+		p.waitForSync(downstream, false)
+	}
+
+	return nil
+}
+
+// initPrimaryDB initializes the local database with the correct users and plugins.
+func (p *Process) initPrimaryDB() error {
+	logger := p.Logger.New("fn", "initPrimaryDB")
+	logger.Info("initializing primary database")
+
+	dsn := p.DSN("127.0.0.1:"+p.Port, "")
+	dsn.User = "root"
+	dsn.Password = ""
+
+	db, err := sql.Open("mysql", dsn.String())
+	if err != nil {
+		logger.Error("error acquiring connection", "err", err)
+		return err
+	}
+	defer db.Close()
+
+	if _, err := db.Exec(fmt.Sprintf(`CREATE USER 'flynn'@'%%' IDENTIFIED BY '%s'`, p.Password)); err != nil {
+		logger.Error("error creating database user", "err", err)
+		return err
+	}
+	if _, err := db.Exec(`GRANT ALL PRIVILEGES ON *.* TO 'flynn'@'%'`); err != nil {
+		logger.Error("error granting privileges", "err", err)
+		return err
+	}
+	if _, err := db.Exec(`GRANT REPLICATION SLAVE ON *.* TO 'flynn'@'%'`); err != nil {
+		logger.Error("error granting replication privileges", "err", err)
+		return err
+	}
+	if _, err := db.Exec(`INSTALL PLUGIN rpl_semi_sync_master SONAME 'semisync_master.so'`); err != nil {
+		logger.Error("error installing rpl_semi_sync_master", "err", err)
+		return err
+	}
+	if _, err := db.Exec(`FLUSH PRIVILEGES`); err != nil {
+		logger.Error("error flushing privileges", "err", err)
+		return err
+	}
+
+	// Enable semi-sync replication on the master.
+	if _, err := db.Exec(`SET GLOBAL rpl_semi_sync_master_enabled = ON`); err != nil {
```

Comment:
We are also going to need to set a few other settings, may actually be better to set them in the configuration file though.

Specifically `rpl_semi_sync_master_wait_no_slave` is required to ensure the timeout will be observed even when the sync isn't connected. Otherwise the primary will automatically switch to async replication and can commit writes that are not replicated anywhere. The timeout itself too needs to be raised to a sufficiently high value that it will never be reached too, which is `rpl_semi_sync_master_timeout`.
We should also ensure that `rpl_semi_sync_master_wait_point` is set to `AFTER_SYNC` which is the safest/most-consistent.